File: src/lib/equipment.ts
/**
 * The kind of slot an item fits — the category, not a specific position. An item
 * declares its `slot` as one of these; a character has concrete named slots
 * ({@link EquipmentSlot}) of each kind.
 */
export const SlotKind = {
  Hand: "hand",
  Finger: "finger",
  Wrist: "wrist",
  Head: "head",
  Torso: "torso",
  Legs: "legs",
  Feet: "feet",
} as const;
export type SlotKind = (typeof SlotKind)[keyof typeof SlotKind];

/**
 * A character's discrete, named, single-occupancy equipment positions. Each holds
 * at most one item (a two-handed weapon spans both hands). Naming each position
 * explicitly — rather than pooling by capacity — lets a future spec remove an
 * individual slot (a lost finger or limb).
 */
export const EquipmentSlot = {
  Head: "head",
  Torso: "torso",
  Legs: "legs",
  Feet: "feet",
  LeftWrist: "leftWrist",
  RightWrist: "rightWrist",
  LeftHand: "leftHand",
  RightHand: "rightHand",
  LeftIndexFinger: "leftIndexFinger",
  LeftRingFinger: "leftRingFinger",
  RightIndexFinger: "rightIndexFinger",
  RightRingFinger: "rightRingFinger",
} as const;
export type EquipmentSlot = (typeof EquipmentSlot)[keyof typeof EquipmentSlot];

/** The kind each named slot belongs to. */
export const SLOT_KIND: Record<EquipmentSlot, SlotKind> = {
  [EquipmentSlot.Head]: SlotKind.Head,
  [EquipmentSlot.Torso]: SlotKind.Torso,
  [EquipmentSlot.Legs]: SlotKind.Legs,
  [EquipmentSlot.Feet]: SlotKind.Feet,
  [EquipmentSlot.LeftWrist]: SlotKind.Wrist,
  [EquipmentSlot.RightWrist]: SlotKind.Wrist,
  [EquipmentSlot.LeftHand]: SlotKind.Hand,
  [EquipmentSlot.RightHand]: SlotKind.Hand,
  [EquipmentSlot.LeftIndexFinger]: SlotKind.Finger,
  [EquipmentSlot.LeftRingFinger]: SlotKind.Finger,
  [EquipmentSlot.RightIndexFinger]: SlotKind.Finger,
  [EquipmentSlot.RightRingFinger]: SlotKind.Finger,
};

/**
 * The default humanoid set of named slots, in canonical fill order (used when
 * `equip` auto-assigns a free slot of a kind).
 */
export const DEFAULT_EQUIPMENT_SLOTS: readonly EquipmentSlot[] = [
  EquipmentSlot.Head,
  EquipmentSlot.Torso,
  EquipmentSlot.Legs,
  EquipmentSlot.Feet,
  EquipmentSlot.LeftWrist,
  EquipmentSlot.RightWrist,
  EquipmentSlot.LeftHand,
  EquipmentSlot.RightHand,
  EquipmentSlot.LeftIndexFinger,
  EquipmentSlot.LeftRingFinger,
  EquipmentSlot.RightIndexFinger,
  EquipmentSlot.RightRingFinger,
];

File: crates/wickedways-core/src/world/equipment.rs
//! Equipment slot constants and lookup helpers.
//! Mirrors.
//!
//! `EquipmentSlot` string constants: the named positions a character can equip items into.
//! `SLOT_KIND`: maps each named slot to its `SlotKind` category.
//! `DEFAULT_EQUIPMENT_SLOTS`: the canonical humanoid fill order.

use crate::world::descriptor::SlotKind;

// Named slot string constants — byte-identical to `EquipmentSlot` values.
pub const HEAD: &str = "head";
pub const TORSO: &str = "torso";
pub const LEGS: &str = "legs";
pub const FEET: &str = "feet";
pub const LEFT_WRIST: &str = "leftWrist";
pub const RIGHT_WRIST: &str = "rightWrist";
pub const LEFT_HAND: &str = "leftHand";
pub const RIGHT_HAND: &str = "rightHand";
pub const LEFT_INDEX_FINGER: &str = "leftIndexFinger";
pub const LEFT_RING_FINGER: &str = "leftRingFinger";
pub const RIGHT_INDEX_FINGER: &str = "rightIndexFinger";
pub const RIGHT_RING_FINGER: &str = "rightRingFinger";

/// The default humanoid slot set in canonical fill order.
/// Mirrors `DEFAULT_EQUIPMENT_SLOTS` exactly.
/// Used by `equip` to assign a free slot in order (first free wins, else displace [0]).
pub const DEFAULT_EQUIPMENT_SLOTS: [&str; 12] = [
    HEAD,
    TORSO,
    LEGS,
    FEET,
    LEFT_WRIST,
    RIGHT_WRIST,
    LEFT_HAND,
    RIGHT_HAND,
    LEFT_INDEX_FINGER,
    LEFT_RING_FINGER,
    RIGHT_INDEX_FINGER,
    RIGHT_RING_FINGER,
];

/// Map a named slot to its `SlotKind` category.
/// Mirrors `SLOT_KIND` record in.
pub fn slot_kind_of(slot: &str) -> Option<SlotKind> {
    match slot {
        HEAD => Some(SlotKind::Head),
        TORSO => Some(SlotKind::Torso),
        LEGS => Some(SlotKind::Legs),
        FEET => Some(SlotKind::Feet),
        LEFT_WRIST | RIGHT_WRIST => Some(SlotKind::Wrist),
        LEFT_HAND | RIGHT_HAND => Some(SlotKind::Hand),
        LEFT_INDEX_FINGER | LEFT_RING_FINGER | RIGHT_INDEX_FINGER | RIGHT_RING_FINGER => {
            Some(SlotKind::Finger)
        }
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::world::descriptor::SlotKind;

    #[test]
    fn default_equipment_slots_has_12_entries() {
        assert_eq!(DEFAULT_EQUIPMENT_SLOTS.len(), 12);
    }

    #[test]
    fn default_equipment_slots_canonical_order() {
        // Mirrors DEFAULT_EQUIPMENT_SLOTS exactly
        assert_eq!(
            DEFAULT_EQUIPMENT_SLOTS,
            [
                "head",
                "torso",
                "legs",
                "feet",
                "leftWrist",
                "rightWrist",
                "leftHand",
                "rightHand",
                "leftIndexFinger",
                "leftRingFinger",
                "rightIndexFinger",
                "rightRingFinger",
            ]
        );
    }

    #[test]
    fn slot_kind_of_maps_all_slots() {
        assert_eq!(slot_kind_of("head"), Some(SlotKind::Head));
        assert_eq!(slot_kind_of("torso"), Some(SlotKind::Torso));
        assert_eq!(slot_kind_of("legs"), Some(SlotKind::Legs));
        assert_eq!(slot_kind_of("feet"), Some(SlotKind::Feet));
        assert_eq!(slot_kind_of("leftWrist"), Some(SlotKind::Wrist));
        assert_eq!(slot_kind_of("rightWrist"), Some(SlotKind::Wrist));
        assert_eq!(slot_kind_of("leftHand"), Some(SlotKind::Hand));
        assert_eq!(slot_kind_of("rightHand"), Some(SlotKind::Hand));
        assert_eq!(slot_kind_of("leftIndexFinger"), Some(SlotKind::Finger));
        assert_eq!(slot_kind_of("leftRingFinger"), Some(SlotKind::Finger));
        assert_eq!(slot_kind_of("rightIndexFinger"), Some(SlotKind::Finger));
        assert_eq!(slot_kind_of("rightRingFinger"), Some(SlotKind::Finger));
    }

    #[test]
    fn slot_kind_of_unknown_returns_none() {
        assert_eq!(slot_kind_of("unknown"), None);
        assert_eq!(slot_kind_of(""), None);
        assert_eq!(slot_kind_of("Hand"), None); // case-sensitive
    }
}
